zqd/ingest/log.go

```diff
@@ -3,79 +3,86 @@ package ingest
 import (
 	"context"
 	"fmt"
-	"io"
 	"os"
-	"path/filepath"
-	"time"
+	"sync/atomic"
 
-	"github.com/brimsec/zq/driver"
 	"github.com/brimsec/zq/pkg/fs"
-	"github.com/brimsec/zq/pkg/nano"
 	"github.com/brimsec/zq/zbuf"
-	"github.com/brimsec/zq/zio"
 	"github.com/brimsec/zq/zio/detector"
-	"github.com/brimsec/zq/zio/zngio"
-	"github.com/brimsec/zq/zng"
 	"github.com/brimsec/zq/zng/resolver"
 	"github.com/brimsec/zq/zqd/api"
-	"github.com/brimsec/zq/zqd/search"
-	"github.com/brimsec/zq/zqd/space"
+	"github.com/brimsec/zq/zqd/storage"
 	"github.com/brimsec/zq/zqe"
-	"github.com/brimsec/zq/zql"
-	"go.uber.org/zap"
 )
 
-const allZngTmpFile = space.AllZngFile + ".tmp"
+// x509.14:00:00-15:00:00.log.gz (open source zeek)
+// x509_20191101_14:00:00-15:00:00+0000.log.gz (corelight)
+const DefaultJSONPathRegexp = `([a-zA-Z0-9_]+)(?:\.|_\d{8}_)\d\d:\d\d:\d\d\-\d\d:\d\d:\d\d(?:[+\-]\d{4})?\.log(?:$|\.gz)`
+
+type LogOp struct {
+	bytesTotal   int64
+	warnings     []string
+	readers      []zbuf.Reader
+	readCounters []*readCounter
+	err          error
+
+	warningCh chan string
+	doneCh    chan struct{}
+}
 
 // Logs ingests the provided list of files into the provided space.
 // Like ingest.Pcap, this overwrites any existing data in the space.
-func Logs(ctx context.Context, pipe *api.JSONPipe, s *space.Space, req api.LogPostRequest) error {
-	ingestDir := s.DataPath(tmpIngestDir)
-	if err := os.Mkdir(ingestDir, 0700); err != nil {
-		// could be in use by pcap or log ingest
-		if os.IsExist(err) {
-			return ErrIngestProcessInFlight
-		}
-		return err
+func NewLogOp(ctx context.Context, store *storage.ZngStorage, req api.LogPostRequest) (*LogOp, error) {
```

Comment:
Interactions with LogOp have 3 different ways to give feedback/warnings/errors:
- an error from NewLogOp
- warnings sent back via channel
- final error sent when done channel closed

Since all the warnings are generated at setup time in NewLogOp, consider changing NewLogOp to:
```go
func NewLogOp(ctx context.Context, store *storage.ZngStorage, req api.LogPostRequest) (op *LogOp, warnings []error, err error)
```
and send all the warnings immediately after the call; that will remove the warnings channel & associated code.

You can also change the done channel to a `chan error`, so the select on it can be:
```go
case err := <-Done()
```

Reply:
@alfred-landrum warnings can also be generated if a reader comes across an invalid record while reading, or runs into a spontaneous error: see line 58

Reply:
w.r.t making Done and `chan error`, the down side to this approach is that only one subscriber will know if an error occurred. If I make the channel a buffered channel there won't even need to be a subscriber in order for the error to be received by consumers.

I suspect this is why the authors of package context took this approach.

Reply:
ok, how about removing the error return from NewLogOp entirely, and having it use a single channel to return warnings on, that when closed, signals the operation is done? So inside log:

```go
  op := NewLogOp(ctx, store, req)
  ...
  select {
  case warning, ok := <- op.Status()
  if !ok {
     // op has finished, get final error if any
     err := op.Error()
     ...
     return
  }
  // Otherwise, send warning:
  err := pipe.Send(api.LogPostWarning{...})
  ...
  }
```

That would reduce the op to use a single channel, and remove the extra code to handle storing & draining the warnings.

Reply:
Yep, I really like that idea.